File: 1743-restore-array.py
def find(pair, num):
    a, b = pair
    if a == num:
        return b
    elif b == num:
        return a
    else:
        return None
    

class Solution:
    def restoreArray(self, adjacentPairs: list[list[int]]) -> list[int]:
        mylist = adjacentPairs.pop()
        l, r = mylist
        i = len(adjacentPairs)-1
        # j = 0
        while adjacentPairs:
            # j += 1
            # if j > 10:
            #     return
            # print(l, mylist, r, adjacentPairs[i], "from", adjacentPairs)
            num = find(adjacentPairs[i], l)
            if num is not None:
                mylist.insert(0, num)
                adjacentPairs.pop(i)
                l = num
                num = None
            else:
                num = find(adjacentPairs[i], r)
                
            if num is not None:
                mylist.append(num)
                r = num
                adjacentPairs.pop(i)
            
            i -= 1
            if i < 0:
                ll = len(adjacentPairs)
                if ll == 0:
                    break
                i %= ll
        
        return mylist
            
        
if __name__ == "__main__":
    ans = Solution().restoreArray([[2,1],[3,4],[3,2]])
    ans = Solution().restoreArray([[4,-2],[1,4],[-3,1]])
    print(ans)

File: 1743-restore-array.rs
fn find(pair: &[i32], num: i32) -> Option<i32> {
    let (a, b) = (pair[0], pair[1]);
    if a == num {
        Some(b)
    } else if b == num {
        Some(a)
    } else {
        None
    }
}

impl Solution {
    fn restore_array(mut adjacent_pairs: Vec<Vec<i32>>) -> Vec<i32> {
        let mut mylist = adjacent_pairs.pop().unwrap();
        let mut l = mylist[0];
        let mut r = mylist[1];
        let mut i : isize = adjacent_pairs.len() as isize -1;
        while !adjacent_pairs.is_empty() {
            {
                let i = i as usize;
                let mut num = find(&adjacent_pairs[i], l);
                if let Some(num) = num {
                    mylist.insert(0, num);
                    adjacent_pairs.remove(i);
                    l = num;
                } else {
                    let mut num = find(&adjacent_pairs[i], r);
                    if let Some(num) = num {
                        mylist.push(num);
                        adjacent_pairs.remove(i);
                        r = num
                    } 
                }
            }
            i -= 1;
            if i < 0 {
                i += adjacent_pairs.len() as isize ;
                if i < 0 {
                    break;
                }
            }
        }
        mylist
    }
}



fn main() {
    let input: Vec<Vec<i32>> = [[-3,-9],[-5,3],[2,-9],[6,-3],[6,1],[5,3],[8,5],[-5,1],[7,2]]
    .iter()
    .map(|s| s.to_vec())
    .collect();
    dbg!(Solution::restore_array(input));
}

struct Solution;